shell/platform/windows/compositor_software.cc

```diff
@@ -51,18 +51,105 @@ bool CompositorSoftware::Present(FlutterViewId view_id,
-  FML_DCHECK(layers[0]->offset.x == 0 && layers[0]->offset.y == 0);
-  FML_DCHECK(layers[0]->type == kFlutterLayerContentTypeBackingStore);
-  FML_DCHECK(layers[0]->backing_store->type ==
-             kFlutterBackingStoreTypeSoftware);
-
-  const auto& backing_store = layers[0]->backing_store->software;
-
-  return view->PresentSoftwareBitmap(
-      backing_store.allocation, backing_store.row_bytes, backing_store.height);
+  // Bypass composition logic if there is only one layer.
+  if (layers_count == 1) {
+    auto& layer = *layers[0];
+    if (layer.type == kFlutterLayerContentTypeBackingStore &&
+        layer.offset.x == 0 && layer.offset.y == 0) {
+      auto& backing_store = *layer.backing_store;
+      FML_DCHECK(backing_store.type == kFlutterBackingStoreTypeSoftware);
+      auto& software = backing_store.software;
+      return view->PresentSoftwareBitmap(software.allocation,
+                                         software.row_bytes, software.height);
+    }
+  }
+
+  // Composite many layers.
+  int x_min = INT_MAX;
+  int x_max = INT_MIN;
+  int y_min = INT_MAX;
+  int y_max = INT_MIN;
+  for (const FlutterLayer** layer = layers; layer < layers + layers_count;
+       layer++) {
+    auto& offset = (*layer)->offset;
+    auto& size = (*layer)->size;
+    x_min = std::min(x_min, static_cast<int>(offset.x));
+    y_min = std::min(y_min, static_cast<int>(offset.y));
+    x_max = std::max(x_max, static_cast<int>(offset.x + size.width));
+    y_max = std::max(y_max, static_cast<int>(offset.y + size.height));
+  }
+
+  int width = x_max - x_min;
+  int height = y_max - y_min;
+  std::vector<uint32_t> allocation(width * height, 0xff000000);
+
+  for (const FlutterLayer** layer = layers; layer < layers + layers_count;
+       layer++) {
+    // TODO(schectman): handle platform view type layers.
+    // https://github.com/flutter/flutter/issues/143375
+    if ((*layer)->type == kFlutterLayerContentTypeBackingStore) {
+      BlendLayer(allocation, **layer, x_min, y_min, width, height);
+    } else {
+      FML_UNREACHABLE();
+      return false;
+    }
+  }
+
+  return view->PresentSoftwareBitmap(static_cast<void*>(allocation.data()),
+                                     width * 4, height);
+}
+
+void CompositorSoftware::BlendLayer(std::vector<uint32_t>& allocation,
+                                    const FlutterLayer& layer,
+                                    int x_min,
+                                    int y_min,
+                                    int width,
+                                    int height) const {
+  FML_DCHECK(layer.type == kFlutterLayerContentTypeBackingStore);
+  auto& backing_store = *layer.backing_store;
+  FML_DCHECK(backing_store.type == kFlutterBackingStoreTypeSoftware);
+  auto src_data =
+      static_cast<const uint32_t*>(backing_store.software.allocation);
+  auto& offset = layer.offset;
+  auto& size = layer.size;
+
+  for (int y_src = 0; y_src < size.height; y_src++) {
```

Comment:
Will take a closer look, but first thoughts:
1. but it _feels_ like it should be possible to update the loop conditions to avoid the two < 0, >= height checks below.
2. this _feels_ like there should be a way to parallelise this a bit

Reply:
Feel free to check now.

Reply:
I suspect the right way to parallelize would be to vectorize this code if possible... But that seems like too much work for an MVP.

Reply:
The only ones I'm familiar with work with HDCs instead of pixel data in memory, and so we would need to create those new DCs per layer to apply them as far as I understand